src/axom/inlet/YAMLReader.cpp

```diff
@@ -0,0 +1,389 @@
+ * The traversal prefers the string-valued name but will attempt to convert to
+ * integer if no such child exists.
+ *******************************************************************************
+ */
+conduit::Node traverseNode(const conduit::Node& root, const std::string& id)
+{
+  if(root.has_path(id))
+  {
+    return root[id];
+  }
+
+  const conduit::Node* node = &root;
+  std::vector<std::string> tokens;
+  axom::utilities::string::split(tokens, id, SCOPE_DELIMITER);
+  for(const auto& token : tokens)
+  {
+    // Prefer the string name, but if it doesn't exist, try converting to int
+    if(node->has_child(token))
+    {
+      node = &((*node)[token]);
+    }
+    else
+    {
+      auto as_int = checkedConvertToInt(token);
+      if(as_int.second && as_int.first < node->number_of_children())
+      {
+        node = &((*node)[as_int.first]);
+      }
+      else
+      {
+        // Bail out and return an empty node to indicate failure
+        return {};
+      }
+    }
+  }
+  return *node;
+}
+
+/*!
+ *******************************************************************************
+ * \brief Copies a Conduit array into an unordered map, using zero-based array
+ * indices as the keys
+ *
+ * \param [in] array The array to copy from
+ * \param [out] map  The map to copy into
+ * \note Implementing to allow for widening/narrowing conversions
+ *******************************************************************************
+ */
+template <typename ConduitType, typename MapValueType>
+void arrayToMap(const conduit::DataArray<ConduitType>& array,
+                std::unordered_map<int, MapValueType>& map)
+{
+  map.clear();
+  for(conduit::index_t i = 0; i < array.number_of_elements(); i++)
+  {
+    // No begin/end iterators are provided by DataArray
+    map[i] = array[i];
+  }
+}
+
+}  // namespace detail
+
+bool YAMLReader::getValue(const conduit::Node& node, int& value)
+{
+  // Match LuaReader functionality - narrow from floating-point
+  if(node.dtype().is_number())
+  {
+    value = node.to_int();
+    return true;
+  }
+  return false;
+}
+
+bool YAMLReader::getValue(const conduit::Node& node, std::string& value)
+{
+  if(node.dtype().is_string())
+  {
+    value = node.as_string();
+    return true;
+  }
+  return false;
+}
+
+bool YAMLReader::getValue(const conduit::Node& node, double& value)
+{
+  // Match LuaReader functionality - promote from integer
+  if(node.dtype().is_number())
+  {
+    value = node.to_float64();
+    return true;
+  }
+  return false;
+}
+bool YAMLReader::getValue(const conduit::Node& node, bool& value)
+{
+  // Boolean literals don't appear to be parsed as such - they are strings
+  if(node.dtype().is_string())
+  {
+    std::string as_str = node.as_string();
+    // YAML 1.2 spec, section 10.3.2
+    // FIXME: Converting the string to lowercase is not strictly correct, it
+    // allows for things like tRue and falsE
+    std::transform(as_str.begin(),
```

Comment:
Could you add a to_lower to the string utilities functions in core?  This is useful in more than here and improves readability.